chapters/unitexpressions.tex

```diff
@@ -89,3 +89,265 @@ \section{The Syntax of Unit Expressions}\label{the-syntax-of-unit-expressions}
+\end{enumerate}
+
+
+\subsection{Unit Inference}\label{unit-inference}
+
+In Modelica unit checking, \firstuse{unit inference} refers to the implicit casting of an expression with empty unit to a corresponding expression with non-empty unit.
+An expression can only get an inferred unit when appearing in a context where the empty unit is not allowed.
+
+When encountering the empty unit in the following situations, the inferred unit is (uniquely) determined by ensuring that unit compatibility requirements are fulfilled:
+\begin{itemize}
+\item
+  In binding equations and modifications:
+  \begin{itemize}
+  \item The entire expression of the binding or modification.
+  \item When the entire expression is an array construction, array concatenation and array range, then apply rules recursively for the direct subexpressions.
+  \end{itemize}
+\item
+  The entire argument expression in a function call.
+% To keep the design close to the bare minimum, this part is currently excluded:
+%\item
+%  For a translation-time constant with value 0.0:
+%  \begin{itemize}
+%  \item Expressions constituting the entire side of an equality or relation.
+%  \item The right hand side of an assignment.
+%  \item The direct subexpressions of array construction, array concatenation, and array range.
+%  \end{itemize}
+\end{itemize}
+
+Otherwise, the inferred unit is \lstinline!"1"!.
+
+
+\subsection{Expressions with Empty Unit}\label{expressions-with-empty-unit}
+
+This section describes conditions under which an expression has empty unit.
+Conditions not given here must not be interpreted as definitely not implying empty unit; instead, the unit may be currently undefined for some expressions, allowing the unit to be properly defined in future versions of the specification.
+
+Basic expressions defined to have empty unit:
+\begin{itemize}
+\item
+  \lstinline!Real! literals.
+\item
+  \lstinline!Integer! expressions implicitly cast to \lstinline!Real!.
+\end{itemize}
+
+Expressions defined to \emph{not} propagate the empty unit, thereby forcing inference of unit \lstinline!"1"!:
+\begin{itemize}
+\item
+  Addition, subtraction, multiplication and division operators when either operand has non-empty unit.
+\item
+  Right operand of binary exponentiation.
+\item
+  Component references outside of functions, where the component's unit (after unit propagation has been carried out) is empty (possibly by not being specified).
+  (Unit checking involving user-defined functions with empty unit on inputs and outputs is currently not defined.)
+\end{itemize}
+
+Built-in non-array operators, functions, and special expressions that propagate any unit (including empty):
+\begin{itemize}
+\item
+  When all operands have the same unit: negation, addition, and subtraction (scalar or element-wise).
+\end{itemize}
+
+Special situations in which the empty unit is propagated up the expression tree:
+\begin{itemize}
+\item
+  Multiplication and division when both operands have empty unit.
```

Comment:
Also addition/subtraction? Or all operators? I'm not sure...

Reply:
Yes, this is where I see a possibility to simplify the presentation by concentrating on just the propagations of the empty unit up the expression tree.  Then all binary operators as well as unary negation would follow this simple rule:
- All binary numeric operators as well as unary negation result in the empty unit if and only if all operands have empty unit.  If one of the operands has non-empty unit, the other operand must have non-empty unit too, meaning that unit inference gets applied in case the other operand's unit is empty.

Reply:
Sounds good.

Reply:
I gave it a try, and it seems to me that this led to considerable simplification of the proposal.  Maybe we could consider this conversation resolved, and continue the discussion in conversations attached to the updated text?

(By the way, I'll be away until Monday.)